package/src/core/define-utility.ts

```diff
@@ -0,0 +1,27 @@
+import type { HookParameters, Hooks } from "./types.js";
+
+/**
+ * Allows defining a type-safe function requiring all the params of a given hook.
+ * It uses currying to make TypeScript happy.
+ *
+ * @param {string} _hook
+ *
+ * @see https://astro-integration-kit.netlify.app/utilities/define-utility/
+ *
+ * @example
+ * ```ts
+ * const test = defineUtility("astro:config:setup")((params, foo: boolean) => {
+ *  return "bar";
+ * });
+ * ```
+ */
+export const defineUtility =
+	<THook extends keyof Hooks>(_hook: THook) =>
+	/**
+	 * The function itself
+	 * @param {Function} fn;
+	 */
+	<TFn extends (params: HookParameters<THook>, ...args: Array<any>) => any>(
```

Comment:
Does this keep typing of `args`? Given that it's type `any` ?

And how does returning a type of `any` work here? Is it just more effort to type as a function? Or is it not needed?

Reply:
since this is a generic type constraint, the type will be automatically inferred! Basically:
- Based on `_hook` value, it infers the `THook` type and the first parameter of the function so it's typed automatically
- Arguments (if any provided) will be inferred correctly
- Same with return type

See https://tsplay.dev/wed7aW

Reply:
It does. The type is `TFn`, so it returns whichever type you pass in. The type is just required to be assignable to this signature, it won't _be_ this signature.

Reply:
Florian beat me to it

Reply:
Jeez, TS is crazy! I feel this is just something you gotta know!

Reply:
@florian-lefebvre @Fryuni Just another one on this one then: For the `args`, why are they typed as `Array<any>` and not `Record<any, any>` ? They should always be passing an object right?

Reply:
The args is a variadic argument. It represents that the function can receive any number of arguments after the params, including none. So, it has to be an array.

It is inferred as the tuple of the following arguments of the function. This means each utility can receive as many arguments as they want and of any type they want.

A function `(params, arg1: string, arg2: number) => {...}` is assignable to that type with the params being inferred from the function call to the function declaration as the type of the hook and the remaining arguments being inferred from the function declaration to the function call as `...args: [arg1: string, arg2: number]` (arrays in TS accept tuples and allow inference of tuples in generics)

See https://tsplay.dev/Wzqprm

Reply:
@Fryuni thank you for that! That makes much more sense and I get it now! Yeah that all makes complete sense. Love it!